### PR TITLE
Refactor proof state

### PR DESCRIPTION
**Pull request by vicsn**
## Motivation

The round functions mutate prover/verifier state and keep passing it as they go along, see as a starting point
algorithms/src/snark/varuna/varuna.rs

This PR cleans that up by making round functions members of prover/verifier `State`. We also make verifier State a member of prover State.

## Test Plan

This is just a refactor which should not impact any values or tests.

### CODE REVIEW

File: algorithms/src/snark/varuna/ahp/prover/round_functions/second.rs [hunk 1]
```diff
@@ -49,44 +42,40 @@ impl<F: PrimeField, MM: SNARKMode> AHPForR1CS<F, MM> {
...
+        Ok(oracles)
     }
 
-    fn calculate_rowcheck_witness(
-        state: &mut prover::State<F, MM>,
-        batch_combiners: &BTreeMap<CircuitId, verifier::BatchCombiners<F>>,
-    ) -> Result<DensePolynomial<F>> {
-        let mut job_pool = ExecutionPool::with_capacity(state.circuit_specific_states.len());
-        let max_constraint_domain = state.max_constraint_domain;
+    fn calc_rowcheck_witness(&mut self) -> Result<DensePolynomial<F>> {
```
Comment (Pratyush) [inline algorithms/src/snark/varuna/ahp/prover/round_functions/second.rs]: I guess we don't want to change these method *names*, right?

File: algorithms/src/snark/varuna/ahp/prover/state.rs [hunk 2]
```diff
@@ -89,9 +93,11 @@ pub struct State<'a, F: PrimeField, MM: SNARKMode> {
     /// The total number of instances we're proving in the batch.
     pub(in crate::snark) total_instances: usize,
     /// Precomputed roots for calculating FFTs
-    pub fft_precomputation: &'a FFTPrecomputation<F>,
+    pub(in crate::snark) fft_precomputation: &'a FFTPrecomputation<F>,
     /// Precomputed inverse roots for calculating inverse FFTs
-    pub ifft_precomputation: &'a IFFTPrecomputation<F>,
+    pub(in crate::snark) ifft_precomputation: &'a IFFTPrecomputation<F>,
+    /// The verifier message in the first round of the AHP
```
Comment (Pratyush) [inline algorithms/src/snark/varuna/ahp/prover/state.rs]: This documentation is incorrect

File: algorithms/src/snark/varuna/varuna.rs [hunk 3]
```diff
@@ -437,15 +433,13 @@ where
             &mut sponge,
         );
 
-        let (verifier_third_msg, verifier_state) =
-            AHPForR1CS::<_, MM>::verifier_third_round(verifier_state, &mut sponge)?;
+        prover_state.verifier_state.as_mut().unwrap().third_round(&mut sponge)?;
```
Comment (Pratyush) [inline algorithms/src/snark/varuna/varuna.rs]: Maybe we can add `ProverState::verifier_nth_round` functions which handle these things?`?

File: algorithms/src/snark/varuna/varuna.rs [hunk 4]
```diff
@@ -549,13 +543,13 @@ where
         }
 
         // Compute the AHP verifier's query set.
-        let (query_set, verifier_state) = AHPForR1CS::<_, MM>::verifier_query_set(verifier_state);
+        let query_set = prover_state.verifier_state.as_ref().unwrap().query_set();
```
Comment (Pratyush) [inline algorithms/src/snark/varuna/varuna.rs]: Ditto: we introduce a `ProverState::query_set()` method

```suggestion
        let query_set = prover_state.query_set();
```

Comment (vicsn) [reply algorithms/src/snark/varuna/varuna.rs]: c894c4bc6

File: algorithms/src/snark/varuna/varuna.rs (hunk 3 above)
Comment (vicsn) [reply algorithms/src/snark/varuna/varuna.rs]: c2ea3e552

File: algorithms/src/snark/varuna/ahp/prover/state.rs (hunk 2 above)
Comment (vicsn) [reply algorithms/src/snark/varuna/ahp/prover/state.rs]: 43e452a7e

File: algorithms/src/snark/varuna/ahp/prover/round_functions/second.rs (hunk 1 above)
Comment (vicsn) [reply algorithms/src/snark/varuna/ahp/prover/round_functions/second.rs]: f7c487afb

**Review by Pratyush — APPROVED** [review]

**Review by ljedrz — APPROVED** [review]
LGTM

**Comment by vicsn** [comment]
Closing this outdated PR for now, unlikely there's bandwidth or interest to review this part of the code any time soon.

**Pull request closed without merging**